### PR TITLE
New updated AppNav

### LINKED ISSUE #1709: Update AppNav
Update the AppNav component to the new design.

### PR DESCRIPTION
**Pull request by SlyBouhafs**
This PR closes #1709.
Here's a link for a demo: http://164.92.183.234:3002/

Old AppNav

Screen Shot 2022-04-10 at 17 31 10
Screen Shot 2022-04-10 at 17 31 30

New AppNav

Screen Shot 2022-04-10 at 17 32 15
Screen Shot 2022-04-10 at 17 31 50
Screen Shot 2022-04-10 at 17 32 02

### Ux changes:
- Highlight the current active navigation link when a user is under the same path.
- Add an Admin badge for admin users.
- Create a new user menu and move the admin menu inside of it.

### CODE REVIEW

File: components/ProfileDropdownMenu.tsx [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+type ProfileDropDownMenuProps = {
+  username: string
+  isAdmin: boolean
+}
+
+type CustomToggleProps = {
+  children?: React.ReactNode
+  className?: string | undefined
+  onClick?: (event: React.MouseEvent<HTMLDivElement, MouseEvent>) => {}
```
Comment (flacial) [inline components/ProfileDropdownMenu.tsx:21]: I think it could be `() => void` because `onClick` returns no value

**Comment by flacial** [comment]
image

AFAIK, the hover effect means this element isn't active (clicked state) and could be clicked and then something would happen or change. When `CURRICULUM` is active or we're in `/curriculum`, I don't think it should show the hover effect. Instead, it should keep showing the active style:

image

File: scss/appNav.module.scss [hunk 2]
```diff
@@ -1,30 +1,92 @@
...
 .nav-item {
-  margin-left: 2em;
-  margin-right: 2em;
+  margin-left: 1em;
+  margin-right: 1em;
+  font-size: 14px;
+  font-weight: 700;
+  color: #565656;
+  text-transform: uppercase;
+  background-color: #fbfbfb;
```
Comment (flacial) [inline scss/appNav.module.scss:26]: NIT: could be hsl

**Comment by flacial** [comment]
If I'm in one of the curriculum's subpages, shouldn't the nav item be active?

image

File: scss/appNav.module.scss [hunk 3]
```diff
@@ -1,30 +1,92 @@
...
+
+.nav-link {
+  color: #565656;
+  font-weight: 700;
+  &:hover {
+    background-color: color.change(
+      variables.$primary,
+      $lightness: 90%
+    ) !important;
+    color: variables.$primary !important;
```
Comment (anthonykhoa) [inline scss/appNav.module.scss:40]: is it necessary to use `!important` in your css changes?

File: components/ProfileDropdownMenu.tsx [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+      key={title}
+      bsPrefix={isActive(path)}
+    >
+      {title}
+    </Dropdown.Item>
+  ))
+
+  const buttonMenuToggle = React.forwardRef(
+    (props: CustomToggleProps, ref: React.Ref<HTMLDivElement>) => (
+      <div
```
Comment (anthonykhoa) [inline components/ProfileDropdownMenu.tsx:71]: Can you try
```
<div
        ref={ref}
        {...props}
      >
        {props.children}
</div>
```

File: components/ProfileDropdownMenu.tsx [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+  const location = router.asPath
+
+  const fullname = _.get(data, 'userInfo.user.name', '')
+
+  const userInfo: UserInfo = {
+    // 'A' stands for Anonymous, in case user did not put in full name
+    username,
+    firstName: fullname.split(' ')[0] || 'A',
+    lastName: fullname.split(' ')[1] || ' ',
+    discordUserId: data?.userInfo?.user?.discordUserId ?? '',
```
Comment (anthonykhoa) [inline components/ProfileDropdownMenu.tsx:42]: Is there a need to check if these fields are ` null` or `undefined` using the nullish coalescing operator? Seems like unneeded additional complexity -- why not just use the logical OR operator? `||`

File: components/ProfileDropdownMenu.tsx [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+  isAdmin
+}) => {
+  const { data } = useUserInfoQuery({
+    variables: { username },
+    skip: !username
+  })
+  const router = useRouter()
+  const location = router.asPath
+
+  const fullname = _.get(data, 'userInfo.user.name', '')
```
Comment (anthonykhoa) [inline components/ProfileDropdownMenu.tsx:35]: Why do you use lodash `get` here, but elsewhere you use optional chaining?

File: components/ProfileDropdownMenu.tsx [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+import styles from '../scss/profileDropDown.module.scss'
+
+type ProfileDropDownMenuProps = {
+  username: string
+  isAdmin: boolean
+}
+
+type CustomToggleProps = {
+  children?: React.ReactNode
+  className?: string | undefined
```
Comment (anthonykhoa) [inline components/ProfileDropdownMenu.tsx:20]: Is the `| undefined` needed? Since you already use the `?` typescript operator to signify its an optional field

File: components/ProfileDropdownMenu.tsx [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+            <>
+              <Dropdown.Header>ADMIN</Dropdown.Header>
+              {adminDropdownMenu}
+              <Dropdown.Divider />
+            </>
+          )}
+          <Dropdown.Item
+            className={`${styles['dropdown-item']} nav-link `}
+            bsPrefix={isActive(`/profile/${username}`)}
+            href={`/profile/${username}`}
```
Comment (anthonykhoa) [inline components/ProfileDropdownMenu.tsx:106]: There should be a constant made for `/profile`, seems like this string is repeated  a lot in this file

File: components/ProfileDropdownMenu.tsx (hunk 1 above)
Comment (anthonykhoa) [reply components/ProfileDropdownMenu.tsx:21]: I think you're right about returning void, but `event` should still be needed as a parameter since it is a mouse event

File: components/ProfileDropdownMenu.tsx (hunk 6 above)
Comment (songz) [reply components/ProfileDropdownMenu.tsx:35]: Yeah let's be consistent.

File: scss/appNav.module.scss (hunk 3 above)
Comment (songz) [reply scss/appNav.module.scss:40]: yeah let's never use `!important`

File: components/AppNav.tsx [hunk 9]
```diff
@@ -39,63 +32,48 @@ const navItems: NavItem[] = [
...
-      </NavLink>
-
-      <LogoutContainer>
-        <div className={`${styles['light-button']} d-inline`}>
-          <Button border ml="2" type="light">
-            Logout
-          </Button>
-        </div>
-      </LogoutContainer>
+    <div>
```
[the remark this answers is not in the document]
Comment (SlyBouhafs) [reply components/AppNav.tsx:62]: No it isn't, i removed it!

File: components/ProfileDropdownMenu.tsx (hunk 7 above)
Comment (SlyBouhafs) [reply components/ProfileDropdownMenu.tsx:20]: No, it isn't really needed! Thanks!

File: components/ProfileDropdownMenu.tsx (hunk 6 above)
Comment (flacial) [reply components/ProfileDropdownMenu.tsx:35]: Use lodash `get` when you want to return a value other than `undefined` when the key is not found. Optional chaining when the `undefined` value is fine.

File: components/ProfileDropdownMenu.tsx (hunk 5 above)
Comment (SlyBouhafs) [reply components/ProfileDropdownMenu.tsx:42]: Yeah, i think you're right about the added complexity but i get typescript errors when i use the `||` operator.
Screen Shot 2022-04-14 at 20 27 10

Comment (songz) [reply components/ProfileDropdownMenu.tsx:42]: maybe `_.get(data, 'userInfo.user.discordUserId', '')`

File: components/ProfileDropdownMenu.tsx (hunk 6 above)
Comment (anthonykhoa) [reply components/ProfileDropdownMenu.tsx:35]: @flacial If we want to display username as empty string instead of undefined, then just add `||`

```
data?.userInfo?.user?.name || ''
```


IMO I prefer to use lodash `get`, but I believe some conversation was had awhile ago that this codebase should use optional chaining instead of `get`. We really shouldn't be mixing the two, its better to be consistent

File: components/ProfileDropdownMenu.tsx (hunk 5 above)
Comment (anthonykhoa) [reply components/ProfileDropdownMenu.tsx:42]: @SlyBouhafs Thats because you're not using optional chaining. If you used optional chaining it should work. Or if you do what @songz suggested it should work

File: components/ProfileDropdownMenu.tsx (hunk 8 above)
Comment (SlyBouhafs) [reply components/ProfileDropdownMenu.tsx:106]: Thanks! Fixed it!

File: components/AppNav.tsx [hunk 10]
```diff
@@ -39,63 +32,48 @@ const navItems: NavItem[] = [
...
-          bsPrefix={styles['dropdown-item']}
-        />
-      )}
     </>
   )
 }
 
-const LoggedInAuthNav: React.FC<{ username: string }> = ({ username }) => {
-  const capitalized = username.charAt(0).toUpperCase() + username.slice(1)
+const LoggedInAuthNav: React.FC<{ username: string; session: any }> = ({
```
Comment (anthonykhoa) [inline components/AppNav.tsx:55]: Please don't use `any` as a type definition when possible. It defeats the purpose of using typescript

File: components/AppNav.tsx [hunk 11]
```diff
@@ -39,63 +32,48 @@ const navItems: NavItem[] = [
...
+  username,
+  session
+}) => {
+  const isAdmin = _.get(session, 'user.isAdmin', false) as boolean
```
Comment (anthonykhoa) [inline components/AppNav.tsx:59]: Is type casting to Boolean necessary?

Comment (SlyBouhafs) [reply components/AppNav.tsx:59]: Indeed, It wasn't necessary, the returned type is already a boolean! Thanks

**Comment by SlyBouhafs** [comment]
> image
> 
> AFAIK, the hover effect means this element isn't active (clicked state) and could be clicked and then something would happen or change. When `CURRICULUM` is active or we're in `/curriculum`, I don't think it should show the hover effect. Instead, it should keep showing the active style:
> 
> image

After some thought, i think we should leave it for accessibility, when using tab to navigate, it feels like a good indicator.

File: scss/appNav.module.scss (hunk 2 above)
Comment (SlyBouhafs) [reply scss/appNav.module.scss:26]: For the colors values, i'll leave them as is and fix them once we setup a final color palette.

File: components/ProfileDropdownMenu.tsx [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+      href={path}
+      key={title}
+      bsPrefix={isActive(path)}
+    >
+      {title}
+    </Dropdown.Item>
+  ))
+
+  const buttonMenuToggle = React.forwardRef(
+    (props: CustomToggleProps, ref: React.Ref<HTMLDivElement>) => (
```
Comment (anthonykhoa) [inline components/ProfileDropdownMenu.tsx:70]: Your CustomToggleProps have fields that are already in `React.FC`. Is the typescript definition for `CustomToggleProps` needed?

Comment (SlyBouhafs) [reply components/ProfileDropdownMenu.tsx:70]: Sorry, i'm not sure what you mean, the `CustomToggleProps` already uses react typescript definitions except for the classname.

**Review by flacial — APPROVED** [review]

File: scss/navLink.module.scss [hunk 13]
```diff
@@ -9,3 +12,8 @@
 .noUnderline {
   text-decoration: none !important;
 }
+
+.active {
```
Comment (SlyBouhafs) [inline scss/navLink.module.scss:16]: Some bootstrap components don't have the bsPrefix or the bsClass so i'm not sure if it's fixable.

File: components/ProfileDropdownMenu.tsx (hunk 12 above)
Comment (anthonykhoa) [reply components/ProfileDropdownMenu.tsx:71]: If you look at the fields in your `customToggleProps`, you'll notice the fields are already attributes of a normal div component, except for the `children`. So there should already be some existing typescript definition out there you can use for this use case. For example an existing typescript definition you used is `React.Ref<HTMLDivElement>` to define the type for ref. There should similarly be one available for div element props

File: yarn.lock [hunk 14]
```diff
@@ -2663,11 +2663,30 @@
     "@sentry/utils" "6.19.6"
     tslib "^1.9.3"
 
+<<<<<<< Updated upstream
 "@sentry/types@6.19.6", "@sentry/types@^6.11.0":
+=======
+<<<<<<< HEAD
+"@sentry/types@6.19.6":
```
Comment (anthonykhoa) [inline yarn.lock:2670]: What package updates were needed in this file? Why were they needed?

File: scss/navLink.module.scss (hunk 13 above)
Comment (anthonykhoa) [reply scss/navLink.module.scss:16]: What proposals did you try out to try to fix the problem of `using !important`? I'm wanting to know what you tried that failed, so that I know not to suggest the methods

File: components/AppNav.tsx [hunk 15]
```diff
@@ -39,72 +32,52 @@ const navItems: NavItem[] = [
...
   return (
     <>
       {navItems.map(button => (
         <NavLink
           {...button}
-          className={`${styles['nav-item']} nav-link`}
+          className={`${styles['nav-item']} ${styles['nav-link']} px-3`}
           key={button.name}
-          activePath={location === button.path}
+          activePath={location.includes(button.path)}
```
Comment (anthonykhoa) [inline components/AppNav.tsx:46]: What was wrong with the previous logic? `location === button.path`

File: components/ProfileDropdownMenu.tsx [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+  username,
+  isAdmin
+}) => {
+  const { data } = useUserInfoQuery({
+    variables: { username },
+    skip: !username
+  })
+  const router = useRouter()
+  const location = router.asPath
+  const PROFILE_PATH = '/profile/' + username
```
Comment (anthonykhoa) [inline components/ProfileDropdownMenu.tsx:34]: profile path should go in a constants file that should be accessible from any part of the codebase. I see a high potential for all `path` values to be re-used in the codebase. For example `/curriculum` appears a lot
https://github.com/garageScript/c0d3-app/search?q=curriculum

File: components/AppNav.tsx [hunk 17]
```diff
@@ -124,15 +97,25 @@ const AppNav: React.FC<{}> = () => {
     // TODO: replace with typing
     const username = _.get(session, 'user.username', '')
 
-    return <LoggedInAuthNav username={username} />
+    return <LoggedInAuthNav username={username} session={session} />
```
Comment (anthonykhoa) [inline components/AppNav.tsx:100]: This component grabs the isAdmin value that `AppNav` component already grabs, and there isn't any other logic other than returning the ProfileDropdownMenu component. What is the benefit of having this `LoggedInAuthNav` component?

File: components/AppNav.tsx (hunk 15 above)
Comment (SlyBouhafs) [reply components/AppNav.tsx:46]: I used `include` to highlight the navigation link when the user is under a subpage of that path, for now the only use case is mostly for the curriculum subpages, to indicate that they are still inside the curriculum path, i believe it's great for the UX. It will also be useful later if we add submenus for the the different new features.

Comment (anthonykhoa) [reply components/AppNav.tsx:46]: Can you please list all UX changes in your PR description

File: yarn.lock (hunk 14 above)
Comment (SlyBouhafs) [reply yarn.lock:2670]: Sorry, i didn't pay full attention to the yarn.lock, i didn't introduce any new deps, that's probably due to my branch not being up to date with master, i'll get it fixed.

File: components/ProfileDropdownMenu.tsx (hunk 12 above)
Comment (SlyBouhafs) [reply components/ProfileDropdownMenu.tsx:71]: Thanks for explanation, i got it fixed.

File: components/AppNav.tsx (hunk 17 above)
Comment (SlyBouhafs) [reply components/AppNav.tsx:100]: You're right about that!

File: components/AppNav.tsx (hunk 15 above)
Comment (SlyBouhafs) [reply components/AppNav.tsx:46]: Description updated!

File: components/ProfileDropdownMenu.tsx (hunk 16 above)
Comment (SlyBouhafs) [reply components/ProfileDropdownMenu.tsx:34]: Right, although the profile path is only used in the profile dropdown menu for now, I feel like this needs to be another different PR, the constants file doesn't exist yet.
I'll reference this issue here so that we can keep track of it: #1185

File: scss/navLink.module.scss (hunk 13 above)
Comment (SlyBouhafs) [reply scss/navLink.module.scss:16]: Alright, i imported the AppNav styles and used them instead because they have higher specificity.

File: components/ProfileDropdownMenu.tsx (hunk 16 above)
Comment (anthonykhoa) [reply components/ProfileDropdownMenu.tsx:28]: So far the only file that uses `/profile` is components/AppNav, but now we are introducing a new file components/ProfileDropdownMenu in this PR that uses it as well. We shouldn't be introducing more tech debt. `components/AppNav` is already an impacted file, so you're really only adding 1 more file to your PR. This current PR is kinda big already though, so I'd suggest to split this PR into smaller ones.

File: components/ProfileDropdownMenu.tsx [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+      href={path}
+      key={title}
+      bsPrefix={isActive(path)}
+    >
+      {title}
+    </Dropdown.Item>
+  ))
+
+  const buttonMenuToggle = React.forwardRef(
+    (props, ref: React.Ref<HTMLDivElement>) => (
```
Comment (anthonykhoa) [inline components/ProfileDropdownMenu.tsx:65]: Why does ref have a type definition, but props doesn't?

Comment (anthonykhoa) [reply components/ProfileDropdownMenu.tsx:65]: Actually we have no unique fields in the props object so this is ok

Comment (anthonykhoa) [reply components/ProfileDropdownMenu.tsx:65]: I think this might throw typescript error saying children doesn't exist in props

Comment (flacial) [reply components/ProfileDropdownMenu.tsx:65]: > by passing it generic types

What do you mean?

Comment (SlyBouhafs) [reply components/ProfileDropdownMenu.tsx:65]: i am not familiar yet with all the type stuff so I used `<HTMLDivElement>` as a type for forwardRef and tsc didn't scream.

File: components/ProfileDropdownMenu.tsx (hunk 16 above)
Comment (SlyBouhafs) [reply components/ProfileDropdownMenu.tsx]: Created Constants for all the paths in AppNav and ProfileDropdownMenu.

File: components/ProfileDropdownMenu.tsx (hunk 18 above)
Comment (anthonykhoa) [reply components/ProfileDropdownMenu.tsx]: This guy had code similar to yours and it ran into issue of children not appearing in props object
https://stackoverflow.com/questions/54654303/using-a-forwardref-component-with-children-in-typescript

In the documentation, it looks like we have to use `props.children` whenever the dropdown component we use contains children
https://react-bootstrap.github.io/components/dropdowns/

I tried it out on a typescript playground too and it gives me the error of children not existing in props
https://codesandbox.io/s/forwardref-typescript-c0b4r?file=/src/App.tsx
Try this code in the `App.tsx` file:
```
import * as React from "react";


const Search = React.forwardRef<HTMLDivElement>((props, ref) => {
  return <div ref={ref} {...props}>{props.children}</div>;
});
Search.displayName = "Search";

export default function App() {
  const input = React.useRef<HTMLInputElement>(null);
  React.useEffect(() => {
    if (input.current) {
      input.current.focus();
    }
  }, []);
  return <Search ref={input}>Banana</Search>;
}
```

I am not sure why the error does it appear on your end, but either way since we are using typescript, we should make sure to have a type definition for all parameters.

File: components/ProfileDropdownMenu.tsx [hunk 19]
```diff
@@ -0,0 +1,126 @@
...
+      key={title}
+      bsPrefix={isActive(path)}
+    >
+      {title}
+    </Dropdown.Item>
+  ))
+
+  const buttonMenuToggle = React.forwardRef<HTMLDivElement>((props, ref) => (
+    <div ref={ref} {...props}>
+      {props.children}
```
Comment (anthonykhoa) [inline components/ProfileDropdownMenu.tsx:67]: we actually dont need to pass props.children since we already do {...props} on line 66

File: yarn.lock [hunk 20]
```diff
@@ -2814,8 +2814,7 @@
     remark-slug "^6.0.0"
     ts-dedent "^2.0.0"
     util-deprecate "^1.0.2"
-
-"@storybook/addon-essentials@^6.4.22":
+  "@storybook/addon-essentials@^6.4.22":
```
Comment (anthonykhoa) [inline yarn.lock:2817]: Can we remove these changes? Looks like only formatting was changed.

Comment (SlyBouhafs) [reply yarn.lock]: Fixed.

File: components/ProfileDropdownMenu.tsx (hunk 18 above)
Comment (SlyBouhafs) [reply components/ProfileDropdownMenu.tsx]: After reading this: https://stackoverflow.com/a/66876771
It is suggested to either create a type with a children prop, use `React.PropsWithChildren` or use `React.ComponentPropsWithoutRef` if we want to forward all the props, i used the later so that it passes all the props.

File: constants/index.js [hunk 21]
```diff
@@ -0,0 +1,10 @@
+// File for all project constant
```
Comment (flacial) [inline constants/index.js:1]: I thought this file will be created in another PR. This PR creates the constants file AND update the AppNav now.

File: components/AppNav.tsx [hunk 22]
```diff
@@ -1,110 +1,85 @@
...
-  { title: 'Lessons', path: '/admin/lessons' },
-  { title: 'Users', path: '/admin/users' },
-  { title: 'Alerts', path: '/admin/alerts' }
-]
 const navItems: NavItem[] = [
-  { path: '/curriculum', name: 'Curriculum' },
+  { path: CURRICULUM_PATH, name: 'Curriculum' },
   {
-    path: 'https://github.com/garageScript/c0d3-app',
+    path: REPO_PATH,
```
Comment (anthonykhoa) [inline components/AppNav.tsx:28]: Looks like this file doesn't use `/profile` anymore, so all these changes to use constants in this file is just tech debt cleanup. Can we move this to another PR? What needs to be prevented is the creation of new tech debt, which the profiledropdown file will introduce without the `/profile` constant. I'd suggest you break this PR into smaller PRs, with one being for profiledropdown menu with constants file.

**Review by songz — APPROVED** [review]

File: components/AppNav.tsx (hunk 22 above)
Comment (songz) [reply components/AppNav.tsx:28]: > I'd suggest you break this PR into smaller PRs, with one being for profiledropdown menu with constants file.

I think @SlyBouhafs can keep this in mind for the future. Current refactoring has no issues I think its good to merge (PRs been open for awhile).

File: components/ProfileDropdownMenu.tsx [hunk 23]
```diff
@@ -0,0 +1,125 @@
...
+              </Button>
+            </div>
+          </LogoutContainer>
+        </Dropdown.Item>
+      </div>
+    </Dropdown>
+  )
+}
+
+export default ProfileDropdownMenu
```
Comment (anthonykhoa) [inline components/ProfileDropdownMenu.tsx:125]: Lets make a story for this component. This will be useful in the future incase anyone needs to use this component again in the future.

Profile component stories are here
https://github.com/garageScript/c0d3-app/tree/master/stories/profile

Comment (anthonykhoa) [reply components/ProfileDropdownMenu.tsx:125]: This PR has been open for awhile so I don't wanna make this a blocker, but please do follow up with making a story for this in another PR since it is a part of engineering practices

Comment (anthonykhoa) [reply components/ProfileDropdownMenu.tsx:125]: This component isn't reusable, so story is not needed

Comment (SlyBouhafs) [reply components/ProfileDropdownMenu.tsx:125]: Marked as resolved because there is no need for a story, the component isn't reusable.

**Review by anthonykhoa — APPROVED** [review]
